### PR TITLE
Add support for `atomic_min*` and `atomic_umin*` intrinsics

### PR DESCRIPTION
**Pull request by adpaco-aws**
### Description of changes: 

Defines a `min` expression on top of other expressions. Then uses the `min` expression to add support for `atomic_min*` and `atomic_umin*` intrinsics.

### Resolved issues:

Part of #1163 

### Testing:

* How is this change tested? Removes 10 tests and adds 2.

* Is this a refactor change? No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: cprover_bindings/src/goto_program/expr.rs [hunk 1]
```diff
@@ -1080,6 +1080,14 @@ impl Expr {
     pub fn r_ok(self, e: Expr) -> Expr {
         self.binop(ROk, e)
     }
+
+    // Expressions defined on top of other expressions
+
+    /// `min(self, e)`
+    pub fn min(self, e: Expr) -> Expr {
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/expr.rs:1087]: If there are side effects, this will evaluate them twice

Comment (adpaco-aws) [reply cprover_bindings/src/goto_program/expr.rs:1118]: Please take a look at the new changes, which complete the definition for `is_side_effect` and assert that neither of these cause side effects.

File: cprover_bindings/src/goto_program/expr.rs [hunk 2]
```diff
@@ -282,13 +282,44 @@ impl Expr {
...
+        /// Returns whether any of the expressions on a vector
+        /// causes side effects or not
+        fn are_side_effect(exprs: &Vec<Expr>) -> bool {
+            for expr in exprs {
+                if expr.is_side_effect() {
+                    return true;
+                }
+            }
+            false
+        }
```
Comment (danielsn) [inline cprover_bindings/src/goto_program/expr.rs:297]: Style ⛏️ : you could use `exprs.iter().any(|e| e.is_side_effect())`

File: tests/kani/Intrinsics/Atomic/Unstable/AtomicMin/main.rs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+// expected result.
+
+#![feature(core_intrinsics)]
+use std::intrinsics::{
+    atomic_min, atomic_min_acq, atomic_min_acqrel, atomic_min_rel, atomic_min_relaxed,
+};
+
+#[kani::proof]
+fn main() {
+    let mut a1 = 1 as u8;
```
Comment (danielsn) [inline tests/kani/Intrinsics/Atomic/Unstable/AtomicMin/main.rs:14]: Is there a reason we only test for `u8`? Could we make this take a `<T>` and then test multiple bitwidths?

File: cprover_bindings/src/goto_program/expr.rs (hunk 2 above)
Comment (adpaco-aws) [reply cprover_bindings/src/goto_program/expr.rs]: Thanks!

File: tests/kani/Intrinsics/Atomic/Unstable/AtomicMin/main.rs (hunk 3 above)
Comment (adpaco-aws) [reply tests/kani/Intrinsics/Atomic/Unstable/AtomicMin/main.rs:14]: Yes, that's pending in #25. But I'd like to do it once all atomic tests are in.

**Review by danielsn — APPROVED** [review]

**Pull request merged**